llvm/tools/sycl-post-link/CUDANativeSpecConstants.cpp

```diff
@@ -0,0 +1,186 @@
+//===- CUDANativeSpecConstants.cpp - CUDA Native Spec Constants Pass ------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// See comments in the header.
+//===----------------------------------------------------------------------===//
+
+#include "CUDANativeSpecConstants.h"
+#include "Support.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/SYCLLowerIR/TargetHelpers.h"
+
+using namespace llvm;
+
+static GlobalVariable *createGlobal(Module &M, size_t Size, StringRef Name) {
+  auto *ArrTy = ArrayType::get(IntegerType::get(M.getContext(), 64), Size);
+  const bool IsConstant = true;
+  const bool ExternallyInitialized = true;
+  const unsigned AS = 4; // ADDRESS_SPACE_CONST
+  return new GlobalVariable(M, ArrTy, IsConstant, GlobalValue::WeakODRLinkage,
+                            Constant::getNullValue(ArrTy), Name, nullptr,
+                            GlobalValue::NotThreadLocal, AS,
+                            ExternallyInitialized);
+}
+
+static unsigned uintFromMDNode(const MDNode *Node, unsigned OpNumber) {
+  auto *CMD =
+      cast<ConstantAsMetadata>(Node->getOperand(OpNumber).get())->getValue();
+  return static_cast<unsigned>(CMD->getUniqueInteger().getZExtValue());
+}
+
+static MDNode *updateMDNode(Module &M, StringRef KernelName) {
+  auto *AnnotationMD = M.getNamedMetadata("nvvm.annotations");
+  assert(AnnotationMD);
+
+  for (auto *MD : AnnotationMD->operands()) {
+    // Kernel entry points are identified using metadata nodes of the form:
+    //   !X = !{<function>, !"kernel", i32 1}
+    auto *Type = dyn_cast<MDString>(MD->getOperand(1));
+    // Only process kernel entry points.
+    if (!Type || Type->getString() != "kernel")
+      continue;
+
+    // Get a pointer to the entry point function from the metadata.
+    const MDOperand &FuncOperand = MD->getOperand(0);
+    if (!FuncOperand)
+      continue;
+    if (auto *FuncConstant = dyn_cast<ConstantAsMetadata>(FuncOperand))
+      if (auto *Func = dyn_cast<Function>(FuncConstant->getValue()))
+        if (Func->getName() == KernelName)
+          return MD;
+  }
+
+  return nullptr;
+}
+
+PreservedAnalyses CUDANativeSpecConstantsPass::run(Module &M,
```

Comment:
This is a huge method (around 120 lines): what about outlining some "steps" into separate functions? I see like three major steps here: create global variables, change kernel signatures, rewrite kernel argument uses.

Reply:
Agreed, will split it to helpers.

Reply:
Done in: 4f91cb850c55941b086733b7d77b482aa3b5f41c